independent-projects/tools/devtools-common/src/main/java/io/quarkus/cli/plugin/PluginMangerState.java

```diff
@@ -173,11 +175,18 @@ public Map<String, Plugin> extensionPlugins() {
         Map<String, Plugin> extensionPlugins = new HashMap<>();
         projectRoot.map(r -> quarkusProject.get()).ifPresent(project -> {
             try {
-                Set<ArtifactKey> installed = project.getExtensionManager().getInstalled().stream()
-                        .map(ArtifactCoords::getKey).collect(Collectors.toSet());
-                for (Extension e : project.getExtensionsCatalog().getExtensions()) {
-                    if (installed.contains(e.getArtifact().getKey())) {
-                        for (String cliPlugin : ExtensionProcessor.getCliPlugins(e)) {
+                Map<ArtifactKey, Extension> allExtensions = new HashMap<>();
+                project.getExtensionsCatalog().getExtensions().forEach(e -> allExtensions.put(e.getArtifact().getKey(), e));
+
+                for (ArtifactCoords artifactCoords : project.getExtensionManager().getInstalled()) {
+                    ArtifactKey artifactKey = artifactCoords.getKey();
+                    Set<ArtifactKey> allKeys = new HashSet<>();
```

Comment:
This could be a list as well, to be more efficient.

In fact, this collection could potentially be avoided if you extracted the body of the loop to its own method (if that's reasonable) then invoke it once for the extension and then iterate through transitives invoking that method.

Reply:
I am using a `Set` to avoid duplicates that could easily occur if there is an overlap in transitives.

Reply:
The set is created per extension, isn't it? In that case there won't be duplicates.

Reply:
touche